core/src/main/scala/cats/MonadFilter.scala

```diff
@@ -4,15 +4,15 @@ import simulacrum.typeclass
 
 /**
  * a Monad equipped with an additional method which allows us to
- * create an "Empty" value for the Monad (for whatever "empty" makes
+ * create an "empty" value for the Monad (for whatever "empty" makes
  * sense for that particular monad). This is of particular interest to
  * us since it allows us to add a `filter` method to a Monad, which is
  * used when pattern matching or using guards in for comprehensions.
  */
-@typeclass trait MonadFilter[F[_]] extends Monad[F] {
+@typeclass trait MonadFilter[F[_]] extends Monad[F] with FunctorFilter[F] {
 
   def empty[A]: F[A]
```

Comment:
> We could just filter everything from any value and get there, no?

That's true, but you would have to have a value to start with. With `Applicative` you have `pure` for that, but with `Functor` this isn't available.

I was thinking about whether the `empty` from `MonadFilter` should be moved to `FunctorFilter` along with the `filter` method at first, but I think this constrains `FunctorFilter` more than we might want. For example, that would mean that the `FunctorFilter` instance for `OptionT[F, A]` would then require an `Applicative[F]` as opposed to the `Functor[F]` that it currently requires. And more generally, I think that what I have in this PR as a `composeFilter` method on `Functor` would need to instead be moved to `Applicative`.

I think that if we were to add an `ApplicativeFilter`, _that_ would be the place to add the `def empty[A]: F[A]`. With `ApplicativeFilter`, you could even implement `empty` and `pure` in terms of each other (with help from `map`).

At this point I haven't added `ApplicativeFilter` because I didn't have a particularly compelling use-case for it and there's already a lot going on in this type class hierarchy. However, I should be able to add it if people have good use-cases for it.